wn/_queries.py

```diff
@@ -330,6 +330,90 @@ def find_entries(
     yield from rows
 
 
+def find_lemmas(
+    forms: Sequence[str] = (),
+    pos: Optional[str] = None,
+    lexicons: Sequence[str] = (),
+    normalized: bool = False,
+    search_all_forms: bool = False,
+) -> Iterator[_Form]:
+    """Find lemmas (canonical forms) matching the given criteria.
+
+    Returns form data for the lemma of each matching entry.
+    """
+    conn = connect()
+    cte = ''
+    params: list = []
+    conditions = []
+
+    if forms:
+        cte = f'WITH wordforms(s) AS (VALUES {_vs(forms)})'
+        or_norm = 'OR normalized_form IN wordforms' if normalized else ''
+        and_rank = '' if search_all_forms else 'AND rank = 0'
+        conditions.append(f'''
+            e.rowid IN
+               (SELECT entry_rowid
+                  FROM forms
+                 WHERE (form IN wordforms {or_norm}) {and_rank})
+        '''.strip())
+        params.extend(forms)
+    if pos:
+        conditions.append('e.pos = ?')
+        params.append(pos)
+    if lexicons:
+        conditions.append(f'lex.id || ":" || lex.version IN ({_qs(lexicons)})')
+        params.extend(lexicons)
+
+    condition = ''
+    if conditions:
+        condition = 'AND ' + '\n           AND '.join(conditions)
+
+    query = f'''
+          {cte}
+        SELECT DISTINCT f.rowid, f.form, f.id, f.script,
+               p.value, p.variety, p.notation, p.phonemic, p.audio,
+               t.tag, t.category
+          FROM forms AS f
+          JOIN entries AS e ON e.rowid = f.entry_rowid
+          JOIN lexicons AS lex ON lex.rowid = e.lexicon_rowid
+          LEFT JOIN pronunciations AS p ON p.form_rowid = f.rowid
+          LEFT JOIN tags AS t ON t.form_rowid = f.rowid
+         WHERE f.rank = 0
+         {condition}
+         ORDER BY f.rowid ASC
+    '''
+
+    # Group results by form_rowid
+    forms_dict: dict[
+        int,
+        tuple[str, Optional[str], Optional[str], list[_Pronunciation], list[_Tag]]
+    ] = {}
+
+    for row in conn.execute(query, params):
+        form_rowid, form, form_id, script = row[0:4]
+        pron_data = row[4:9]
+        tag_data = row[9:11]
+
+        if form_rowid not in forms_dict:
+            forms_dict[form_rowid] = (form, form_id, script, [], [])
+
+        # Add pronunciation if present
+        if pron_data[0] is not None:  # value
+            pron = cast(_Pronunciation, pron_data)
+            if pron not in forms_dict[form_rowid][3]:
+                forms_dict[form_rowid][3].append(pron)
+
+        # Add tag if present
+        if tag_data[0] is not None:  # tag
+            tag = cast(_Tag, tag_data)
+            if tag not in forms_dict[form_rowid][4]:
+                forms_dict[form_rowid][4].append(tag)
+
+    # Yield forms in order
+    for form_rowid in sorted(forms_dict.keys()):
+        yield forms_dict[form_rowid]
```

Comment:
My previous comment suggested reusing `find_entries()` and `get_entry_forms()`, but if not, I think this sorting is unnecessary. Dictionaries are ordered in Python now, and the SQL query already ordered the results by the form's rowid, so the dictionary should already be in the expected order.

Reply:
Just accepted the solution on mobile. I don't know if it looks quite right but I won't be home for a while. 😅